src/m_sun.cc

```diff
@@ -104,26 +107,110 @@ void sunsAddSingleFromGrid(ArrayOfSun &suns,
 
-  // interpolate field
-  Matrix int_data = regrid_sun_spectrum(sun_spectrum_raw, f_grid, stokes_dim, temperature, verbosity);
+  // init sun
+  Sun& new_sun = suns.emplace_back();
+  new_sun.spectrum = regrid_sun_spectrum(sun_spectrum_raw, f_grid, stokes_dim, temperature, verbosity); // set spectrum
+  new_sun.description = description;
+  new_sun.radius = radius;
+  new_sun.distance = distance;
+  new_sun.latitude = latitude;
+  new_sun.longitude = longitude;
+
+  // set flag
+  suns_do = 1;
+
+}
+
+/* Workspace method: Doxygen documentation will be auto-generated */
+void sunsAddSingleFromGridAtLocation(
+                         ArrayOfSun &suns,
+                         Index &suns_do,
+                         // Inputs:
+                         const Vector &f_grid,
+                         const Index &stokes_dim,
+                         const Vector &refellipsoid,
+                         const GriddedField2 &sun_spectrum_raw,
+                         const Numeric &radius,
+                         const Numeric &distance,
+                         const Numeric &temperature,
+                         const Numeric &zenith,
+                         const Numeric &azimuth,
+                         const String &description,
+                         const Numeric &location_latitude,
+                         const Numeric &location_longitude,
+                         const Numeric &location_altitude,
+                         const Verbosity &verbosity) {
 
-  // create sun
-  Sun& new_star = suns.emplace_back();
+  // some sanity checks
+  ARTS_USER_ERROR_IF (distance<radius,
+                      "The distance to the center of the sun (",distance," m) \n"
+                      "is smaller than the radius of the sun (", radius," m )")
+  ARTS_USER_ERROR_IF (location_altitude<0.,
+                      "The altitude of the solar spectrum should be positiv,\n"
+                      "but is ",location_altitude," m) ")
+
+  // from local position to global position
+  Numeric toa_altitude = location_altitude + refell2r(refellipsoid, location_latitude);
+  
+  Numeric sun_altitude, sun_latitude, sun_longitude;
+  if (zenith < ANGTOL){
+    sun_altitude = distance + toa_altitude;
+    sun_latitude = location_latitude;
+    sun_longitude = location_longitude;
+  } else if (zenith > 180 - ANGTOL) {
+    sun_altitude = distance - toa_altitude;
+    sun_latitude = -location_latitude;
+    sun_longitude = location_longitude + 180 - 360.0 * Numeric(round((location_longitude - 0.0) / 360.0));
+  } else {
+  Numeric x, y, z, dx, dy, dz;
+    poslos2cart(x,
+                y,
+                z,
+                dx,
+                dy,
+                dz,
+                toa_altitude,
+                location_latitude,
+                location_longitude,
+                zenith,
+                azimuth);
+
+    Vector sun_pos_cart = {x+distance*dx, y+distance*dy, z+distance*dz};
```

Comment:
It is not necessary to create this Vector

Reply:
I think it is not the best for readability to remove the Vector, however, you are right, performance wise it is unnecessary.

Reply:
Then call it `std::array<Numeric, 3>` instead.  That is about as good as it gets.

Reply:
I'm fine with it, because it can still be figured out quickly. Anyhow it is good to know how it would be done the best way.